resolve/handle Connection Error on upload

2025-02-09 21:38:09 [bridge      ] [WARNI]  Ignoring broadcast due to invalid temperature: 71.9F
2025-02-09 21:38:13 [bridge      ] [WARNI]  Ignoring broadcast due to invalid temperature: 74.1F
2025-02-09 21:38:14 [PvdrTimer   ] [DEBUG]  timer reset to default
2025-02-09 21:38:14 [bridge      ] [DEBUG]  upload due for Grainfather Custom URL
2025-02-09 21:38:14 [TiltHistory ] [DEBUG]  averaging period set, get average 300 secs
2025-02-09 21:38:14 [TiltHistory ] [DEBUG]  53 averaged uncal values, temp;73.00999 SG:1.0217
2025-02-09 21:38:15 [GFcstm_pvdr ] [ERROR]  ConnectionError: uploading Grainfather Custom device
2025-02-09 21:38:15 [GFcstm_pvdr ] [ERROR]  exception in provider.update: requests Connection error.
2025-02-09 21:38:15 [bridge      ] [CRITI]  handler err: 'NoneType' object isn't iterable
2025-02-09 21:38:15 [bridge      ] [INFO ]  Error in bridge_main: 'NoneType' object isn't iterable
...stopped: Tilt Scanner ('NoneType' object isn't iterable)

--- a/bridge/bridge_main_averaging.py
+++ b/bridge/bridge_main_averaging.py
@@ -238,34 +238,21 @@ async def _beacon_callback(uuid, major, minor, tx_power, rssi, simulated):#, bri
         elif not beacon_data.gravity_valid:
             logger.warning("Ignoring broadcast due to invalid gravity: " + str(beacon_data.gravity))
         else:
-            #logger.info("debug: putting...\n {}".format(dir(beacon_data)))
-            #bridge_q.put_nowait(beacon_data)
-            #bridge_q.put_sync(beacon_data) #, block=False) # Raises IndexError if the queue is full
-            
-            # todo show these messages only if in cal mode - maybe a button, or different main.py & timer
-            # in cal mode perhaps show 30 values then show the average, repeat
-            '''if debug_recvd_counter < 30:
-                # only print to screen TODO this seems to consume memory e.g. increase >60 gc here?
-                ###print(f"data: {colour} SG:{beacon_data.gravity} {beacon_data.temp_fahrenheit}°F")
-                pass
-            debug_recvd_counter += 1
-            '''
-            print(f"data: {colour} SG:{beacon_data.gravity} {beacon_data.temp_fahrenheit}°F")
+            if data_archive.print_raw:
+                # check if we should print raw values as they are received (useful for calibration)
+                print(f"data: {colour} SG:{beacon_data.gravity} {beacon_data.temp_fahrenheit}°F")
             
             try:
                 #await bridge_q.put(beacon_data)
-                # add raw to data archive (for size, storing integer values for Temp & Gravity 1040, not 1.040 or calibrated vals))
+                # add raw to data archive (for size, storing integer values for Temp & Gravity 1040, not 1.040 not calibrated vals))
                 data_archive.add_data(colour, major, minor, time.time())
                 #logger.info(f"added:{colour}, {major}, {minor}, {time.time()}")
-                #data_archive.add_data(colour, sg=1200, tempF=55, tstamp=1724432992)
             except Exception as e:
                 logger.error(f"queue put error: {e}")
                 raise
             #logger.info("{}\t beacon packet received".format(beacon_data.timestamp))
-            #logger.info("debug: bridge_q after {}".format(bridge_q.qsize()))
-        #logger.info("debug: end of if colour")
     else:
-        # todo enable this only if simulated !+ true
+        # if simulated !+ true then warn about unconfigured Tilt
         if simulated == False:
             logger.warning(f"data received for an unconfigured Tilt: {colour}")
         #pass

--- a/bridge/bridge_main_averaging.py
+++ b/bridge/bridge_main_averaging.py
@@ -23,7 +23,7 @@ from configuration import BridgeConfig
 from models.provider_timer import UploadTimers
 gc.collect()
 
-debug_recvd_counter = 0  #for dev purposes, need a better implementation
+#debug_recvd_counter = 0  #for dev purposes, need a better implementation
 
 logger = logging.getLogger('bridge')
 '''
@@ -196,7 +196,7 @@ async def _scan_for_ibeacons(simulate=False):
                         logger.info(f"MAC: {result.device.addr_hex()} Beacon: {result.name}")
 
                         # Call the callback function with the extracted data
-                        await _beacon_callback(uuid, major, minor, tx_power, rssi)#, bridge_q)
+                        await _beacon_callback(uuid, major, minor, tx_power, rssi, simulate)#, bridge_q)
                         #_beacon_callback("a495bb40-c5b1-4b44-b512-1370f02d74df", 65, 1021, 0, 0, bridge_q)
                     if simulate:
                         #logger.info(f"MAC: {result.device.addr_hex()} Not beacon: {result.rssi}")
@@ -205,7 +205,7 @@ async def _scan_for_ibeacons(simulate=False):
                         uuid = random.choice(list(uuid_to_colours.keys()))
                         major = random.randrange(700, 750) # HD ->SD (50, 85)
                         minor = random.randrange(10150, 10350) # HD -> SD (1005, 1045)
-                        await _beacon_callback(uuid, major, minor, 0, 0)#, bridge_q)
+                        await _beacon_callback(uuid, major, minor, 0, 0, simulate)#, bridge_q)
                         #pass # testing is it scanner or callback that causes issue? or maybe colours_to_uuid def?
             except AttributeError:
                 #logger.info(f"scanner result is:{result} scanner is:{scanner}")
@@ -216,12 +216,12 @@ async def _scan_for_ibeacons(simulate=False):
 
 
 #def _beacon_callback(bt_addr, rssi, packet, additional_info, bridge_q):
-async def _beacon_callback(uuid, major, minor, tx_power, rssi):#, bridge_q):
+async def _beacon_callback(uuid, major, minor, tx_power, rssi, simulated):#, bridge_q):
     global data_archive
     # todo: this isn't actually an async routine
     # check bluetooth data and store on a queue (TiltHistory object)
     #    return
-    global debug_recvd_counter
+    #global debug_recvd_counter
     colour = uuid_to_colours.get(uuid)
     if colour in data_archive.ringbuffer_list:
         #logger.info("beacon_callback colour match, {}".format(colour))
@@ -229,6 +229,7 @@ async def _beacon_callback(uuid, major, minor, tx_power, rssi):#, bridge_q):
         # major = degrees in F (int)
         # minor = gravity (int) - needs to be converted to float (e.g. 1035 -> 1.035)
         #start = gc.mem_free()
+        gc.collect() #testing
         beacon_data = TiltStatus(colour, major, _get_decimal_gravity(minor), config)
         #logger.info("cb_tilt_status is:{} bytes".format(start - gc.mem_free()))
         #logger.info("debug: tilt_status:\n{}".format(dir(tilt_status)))
@@ -243,10 +244,13 @@ async def _beacon_callback(uuid, major, minor, tx_power, rssi):#, bridge_q):
             
             # todo show these messages only if in cal mode - maybe a button, or different main.py & timer
             # in cal mode perhaps show 30 values then show the average, repeat
-            if debug_recvd_counter < 60:
-                # only print to screen
-                print(f"data from {colour} tilt SG:{beacon_data.gravity} {beacon_data.temp_fahrenheit}°F")
+            '''if debug_recvd_counter < 30:
+                # only print to screen TODO this seems to consume memory e.g. increase >60 gc here?
+                ###print(f"data: {colour} SG:{beacon_data.gravity} {beacon_data.temp_fahrenheit}°F")
+                pass
             debug_recvd_counter += 1
+            '''
+            print(f"data: {colour} SG:{beacon_data.gravity} {beacon_data.temp_fahrenheit}°F")
             
             try:
                 #await bridge_q.put(beacon_data)
@@ -261,9 +265,11 @@ async def _beacon_callback(uuid, major, minor, tx_power, rssi):#, bridge_q):
             #logger.info("debug: bridge_q after {}".format(bridge_q.qsize()))
         #logger.info("debug: end of if colour")
     else:
-        #logger.info(f"beacon_callback no colour match: {colour}")
-        #todo: log a warning here
-        pass
+        # todo enable this only if simulated !+ true
+        if simulated == False:
+            logger.warning(f"data received for an unconfigured Tilt: {colour}")
+        #pass
+
 
 
 #def _handle_bridge_queue(enabled_providers: list, console_log: bool):

--- a/bridge/lib/models/tilt_history.py
+++ b/bridge/lib/models/tilt_history.py
@@ -17,6 +17,7 @@
     todo: check all necessary defs are present - see primitives/ringbuffer_queue.py
     todo: check what happens if we pass data from a TiltPro in (4 decimcal places)
 '''
+from machine import Timer
 from configuration import BridgeConfig
 #from .json_serialize import JsonSerialize
 import time
@@ -30,7 +31,9 @@ logger = logging.getLogger('TiltHistory')
 
 
 class TiltHistory():
-    # 
+    ''' stores received raw (uncalibrated) sample data for each Tilt device
+        in a binary format, in a circular buffer
+    '''
     def __init__(self, colour_dict):
         # colour dict is; colour: max of averaging period
         #if kwargs:
@@ -40,8 +43,16 @@ class TiltHistory():
         #self.data_points = config.averaging_period # todo allow this per provider ...
         # for each colour in config find max averaging
         # get a list of colour:number
+        self.print_raw = True
+        self.print_timer = Timer(
+            mode=Timer.ONE_SHOT, period=15_000, callback=self._timeout_callback
+        )
         self.ringbuffer_list = dict()
         self.initialise_ringbuffer(colour_dict) # create appropriately sized buffer(s) #todo: colour_dict
+        
+    def _timeout_callback(self, timer):
+        # stop printing data: statements to serial
+        self.print_raw = False
         
     def initialise_ringbuffer(self, colour_dict): 
         # create empty buffer(s)
@@ -238,9 +249,10 @@ class TiltRingBuffer:
             self._ri = (self._ri + c) % self._size  # Discard a message
             raise IndexError  # Caller can ignore if overwrites are OK
 
-    async def _put(self, data):  # Usage: await queue.put(item)
+    '''async def _put(self, data):  # Usage: await queue.put(item)
         # TODO #7 is this function used, possibly left over, otherwise add attribute TiltRungBuffer.full
         while self.full():  # Queue full
             await self._evget.wait()  # May be >1 task waiting on ._evget
             # Task(s) waiting to get from queue, schedule first Task
-        self._put_nowait(data)   
+        self._put_nowait(data)
+    '''

